apstools/_devices/positioner_soft_done.py

```diff
@@ -0,0 +1,144 @@
+        Disregarded if readback attribute is created.
+    setpoint_pv : str, optional
+        PV prefix of the setpoint signal.
+        Disregarded if setpoint attribute is created.
+    tolerance : float, optional
+        Motion tolerance. The motion is considered *done* when::
+
+            abs(readback-setpoint) <= tolerance
+
+        Defaults to ``10^(-1*precision)``,
+        where ``precision = setpoint.precision``.
+    target_attr : str, optional
+        Used if the setpoint is controlled incrementally by EPICS
+        (like with a ramp). Then a target attribute signal must be
+        defined, and its name passed in this variable.
+    kwargs :
+        Passed to `ophyd.PVPositioner`
+
+    ATTRIBUTES
+
+    setpoint : Signal
+        The setpoint (request) signal
+    readback : Signal or None
+        The readback PV (e.g., encoder position PV)
+    actuate : Signal or None
+        The actuation PV to set when movement is requested
+    actuate_value : any, optional
+        The actuation value, sent to the actuate signal when motion is
+        requested
+    stop_signal : Signal or None
+        The stop PV to set when motion should be stopped
+    stop_value : any, optional
+        The value sent to stop_signal when a stop is requested
+
+    (new in apstools 1.5.2)
+    """
+
+    # positioner
+    readback = FormattedComponent(
+        EpicsSignalRO, "{prefix}{_readback_pv}", kind="hinted", auto_monitor=True
+    )
+    setpoint = FormattedComponent(
+        EpicsSignal, "{prefix}{_setpoint_pv}", kind="normal", put_complete=True
+    )
+    done = Component(Signal, value=True, kind="config")
+    done_value = True
+
+    tolerance = Component(Signal, value=-1, kind="config")
+    report_dmov_changes = Component(Signal, value=False, kind="omitted")
+
+    @property
+    def precision(self):
+        return self.setpoint.precision
+
+    def cb_readback(self, *args, **kwargs):
+        """
+        Called when readback changes (EPICS CA monitor event).
+        """
+        diff = self.readback.get() - self._target.get()
+        _tolerance = (
+            self.tolerance.get()
+            if self.tolerance.get() >= 0
+            else 10 ** (-1 * self.precision)
+        )
+        dmov = abs(diff) <= _tolerance
+        if self.report_dmov_changes.get() and dmov != self.done.get():
+            logger.debug(f"{self.name} reached: {dmov}")
+        self.done.put(dmov)
+
+    def cb_setpoint(self, *args, **kwargs):
+        """
+        Called when setpoint changes (EPICS CA monitor event).
+        When the setpoint is changed, force done=False.  For any move,
+        done must go != done_value, then back to done_value (True).
+        Without this response, a small move (within tolerance) will not return.
+        Next update of readback will compute self.done.
+        """
+        self.done.put(not self.done_value)
+
+    def __init__(
+        self,
+        prefix="",
+        *,
+        readback_pv="",
```

Comment:
Wouldn't you want to trap for the case when `readback_pv == setpoint_pv`?  This device would be totally unnecessary in that case.  Better to use `EpicsSignal` instead, in that case.

Reply:
Just by raising an error if that is the case? Or do you have something fancier in mind?